File: server/internal/business/models/category.go
package models

import (
	"database/sql/driver"
	"fmt"
	"time"

	"github.com/google/uuid"
)

type CategoryType string

const (
	CategoryTypeIncome  CategoryType = "income"
	CategoryTypeExpense CategoryType = "expense"
)

type Category struct {
	ID             uuid.UUID     `db:"id" json:"id"`
	UserID         uuid.UUID     `db:"user_id" json:"user_id"`
	Name           string        `db:"name" json:"name"`
	Type           CategoryType  `db:"type" json:"type"`
	ParentID       *uuid.UUID    `db:"parent_id" json:"parent_id"`
	Icon           string        `db:"icon" json:"icon"`
	CreatedAt      time.Time     `db:"created_at" json:"created_at"`
	UpdatedAt      time.Time     `db:"updated_at" json:"updated_at"`
	LastModifiedAt time.Time     `db:"last_modified_at" json:"last_modified_at"`
	Version        int           `db:"version" json:"version"`
	IsDeleted      bool          `db:"is_deleted" json:"is_deleted"`
}

func (c *CategoryType) Scan(value interface{}) error {
	str, ok := value.(string)
	if !ok {
		return fmt.Errorf("invalid category type: %v", value)
	}
	*c = CategoryType(str)
	return nil
}

func (c CategoryType) Value() (driver.Value, error) {
	return string(c), nil
}

// Default categories
var DefaultIncomeCategories = []string{
	"Salary", "Bonus", "Investment", "Gift", "Refund", "Reimbursement", "Other Income",
}

var DefaultExpenseCategories = []string{
	"Food & Dining", "Groceries", "Transportation", "Shopping", "Entertainment",
	"Healthcare", "Education", "Utilities", "Rent", "Travel", "Subscriptions",
	"Personal Care", "Gifts", "Other Expense",
}

File: flutter/lib/data/models/category.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

import '../../sync/lww_strategy.dart';

part 'category.g.dart';

enum CategoryType {
  @JsonValue('income')
  income,
  @JsonValue('expense')
  expense,
}

@JsonSerializable()
class Category extends Equatable implements LwwEntity {
  final String id;
  @JsonKey(name: 'user_id')
  final String userId;
  final String name;
  final CategoryType type;
  @JsonKey(name: 'parent_id')
  final String? parentId;
  final String? icon;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;
  @JsonKey(name: 'last_modified_at')
  final DateTime lastModifiedAt;
  final int version;
  @JsonKey(name: 'is_deleted')
  final bool isDeleted;

  const Category({
    required this.id,
    required this.userId,
    required this.name,
    required this.type,
    this.parentId,
    this.icon,
    required this.createdAt,
    required this.updatedAt,
    required this.lastModifiedAt,
    required this.version,
    required this.isDeleted,
  });

  Category copyWith({
    String? id,
    String? userId,
    String? name,
    CategoryType? type,
    String? parentId,
    String? icon,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? lastModifiedAt,
    int? version,
    bool? isDeleted,
  }) {
    return Category(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      name: name ?? this.name,
      type: type ?? this.type,
      parentId: parentId ?? this.parentId,
      icon: icon ?? this.icon,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      lastModifiedAt: lastModifiedAt ?? this.lastModifiedAt,
      version: version ?? this.version,
      isDeleted: isDeleted ?? this.isDeleted,
    );
  }

  factory Category.fromJson(Map<String, dynamic> json) => _$CategoryFromJson(json);

  Map<String, dynamic> toJson() => _$CategoryToJson(this);

  @override
  List<Object?> get props => [
    id, userId, name, type, parentId, icon, createdAt,
    updatedAt, lastModifiedAt, version, isDeleted,
  ];
}
